opm/simulators/wells/WellConvergence.cpp

```diff
@@ -0,0 +1,155 @@
+checkConvergenceControlEq(const WellState& well_state,
+                          const Tolerances& tolerances,
+                          const double well_control_residual,
+                          ConvergenceReport& report,
+                          DeferredLogger& deferred_logger) const
+{
+    double control_tolerance = 0.;
+    using CR = ConvergenceReport;
+    CR::WellFailure::Type ctrltype = CR::WellFailure::Type::Invalid;
+
+    const int well_index = well_.indexOfWell();
+    const auto& ws = well_state.well(well_index);
+    if (well_.wellIsStopped()) {
+        ctrltype = CR::WellFailure::Type::ControlRate;
+        control_tolerance = tolerances.rates; // use smaller tolerance for zero control?
+    }
+    else if (well_.isInjector() )
+    {
+        auto current = ws.injection_cmode;
+        switch(current) {
+        case Well::InjectorCMode::THP:
+            ctrltype = CR::WellFailure::Type::ControlTHP;
+            control_tolerance = tolerances.thp;
+            break;
+        case Well::InjectorCMode::BHP:
+            ctrltype = CR::WellFailure::Type::ControlBHP;
+            control_tolerance = tolerances.bhp;
+            break;
+        case Well::InjectorCMode::RATE:
+        case Well::InjectorCMode::RESV:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.rates;
+            break;
+        case Well::InjectorCMode::GRUP:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.grup;
+            break;
+        default:
+            OPM_DEFLOG_THROW(std::runtime_error, "Unknown well control control types for well " << well_.name(), deferred_logger);
+        }
+    }
+    else if (well_.isProducer() )
+    {
+        auto current = ws.production_cmode;
+        switch(current) {
+        case Well::ProducerCMode::THP:
+            ctrltype = CR::WellFailure::Type::ControlTHP;
+            control_tolerance = tolerances.thp;
+            break;
+        case Well::ProducerCMode::BHP:
+            ctrltype = CR::WellFailure::Type::ControlBHP;
+            control_tolerance = tolerances.bhp;
+            break;
+        case Well::ProducerCMode::ORAT:
+        case Well::ProducerCMode::WRAT:
+        case Well::ProducerCMode::GRAT:
+        case Well::ProducerCMode::LRAT:
+        case Well::ProducerCMode::RESV:
+        case Well::ProducerCMode::CRAT:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.rates;
+            break;
+        case Well::ProducerCMode::GRUP:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.grup;
+            break;
+        default:
+            OPM_DEFLOG_THROW(std::runtime_error, "Unknown well control control types for well " << well_.name(), deferred_logger);
+        }
+    }
+
+    const int dummy_component = -1;
+    if (std::isnan(well_control_residual)) {
```

Comment:
We must include `<cmath>` in order to use `std::isnan()`.

Reply:
done.